### PR TITLE
Clean up XHR API

### PR DESCRIPTION
**Pull request by KiChjang**
I've also added annotations about the steps that we're performing within each method.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9383)

### CODE REVIEW

File: components/script/dom/xmlhttprequest.rs [hunk 1]
```diff
@@ -750,20 +791,19 @@ impl XMLHttpRequestMethods for XMLHttpRequest {
...
+                        let response = self.document_response();
+                        self.response_xml.set(response.r());
+                        response
+                    }))
+                } else {
+                // Step 2
+                    Ok(None)
                 }
             },
+            // Step 1
```
Comment (frewsxcv) [inline components/script/dom/xmlhttprequest.rs]: Move this above since it is step 1

Comment (frewsxcv) [inline components/script/dom/xmlhttprequest.rs]: Nevermind, not trivial since it's matching on `_`

**Comment by Ms2ger** [comment]
How about a `sync_allowed` method that checks the global and the sync flag?

File: components/script/dom/xmlhttprequest.rs [hunk 2]
```diff
@@ -295,6 +295,9 @@ impl XMLHttpRequestMethods for XMLHttpRequest {
 
     // https://xhr.spec.whatwg.org/#the-open()-method
     fn Open(&self, method: ByteString, url: USVString) -> ErrorResult {
+        // Step 1
+        if !self.global().r().as_window().Document().r().is_fully_active() { return Err(Error::InvalidState); }
```
Comment (jdm) [inline components/script/dom/xmlhttprequest.rs]: This will cause a panic when calling this method in a worker.

Comment (KiChjang) [inline components/script/dom/xmlhttprequest.rs]: > If settings object's responsible document is not fully active, throw an InvalidStateError exception.

Hmm... how would we get a `responsible document` from a `Worker`, in that case?

Comment (jdm) [inline components/script/dom/xmlhttprequest.rs]: That concept is not implemented yet. We also will not be able to interact with the `Document` on the worker thread; we would need to block on a runnable sent to the document's associated script thread.

**Comment by KiChjang** [comment]
@Ms2ger Done.

**Comment by Ms2ger** [comment]
Looking at this again, `sync_allowed` doesn't really make sense as a method name. Maybe `sync_in_window`? Feels like we should be able to do better.

**Comment by KiChjang** [comment]
Done that as well.

**Comment by Ms2ger** [comment]
You made me read through most of the XHR spec, so now you get to file issues for all the bugs I found :)

-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 197 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fi4uMGujeO-0Hpwjy:-K9fi4uOTbXjTmecydUU:2061635503) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L197)):*
Now the value you're returning is backwards :)

---

*[components/script/dom/xmlhttprequest.rs, line 308 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fkeam9hi9GbqUTzUC:-K9fkeanAZzGVoVqaXyL:1938472062) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L308)):*
Not entirely sure this is useful, since it boils down to `browsing_context.active_document() == browsing_context.active_document()` right now.

---

*[components/script/dom/xmlhttprequest.rs, line 310 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fiiZkj0d-a8pmF_B_:-K9fiiZmvIb4TDSl-oQZ:1714631263) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L310)):*
This issue has been fixed; remove the reference.

---

*[components/script/dom/xmlhttprequest.rs, line 329 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fl1Jmw6AG_Rlkq0ep:-K9fl1Jmw6AG_Rlkq0eq:-1806053217) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L329)):*
This is no longer step 2.

---

*[components/script/dom/xmlhttprequest.rs, line 347 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fptpKELotq-DoPz-T:-K9fptpMSzN5xRsLtptS:-1796807218) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L347)):*
Please file an issue and point to it here.

---

*[components/script/dom/xmlhttprequest.rs, line 388 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fswTNN0qMCovx4_KA:-K9fswTPbOCtpKJolkIT:-1842496409) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L388)):*
We don't seem to do step 3; please file an issue and add a comment here.

---

*[components/script/dom/xmlhttprequest.rs, line 416 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fuOpBiiIf0YjTkdU4:-K9fuOpDE7O5QoXpTFhI:-1985273620) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L416)):*
What step is this? Is it reachable? If so, do we have tests that reach it?

---

*[components/script/dom/xmlhttprequest.rs, line 436 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fwCzbNZnmQK79yG4K:-K9fwCzdpNokH6Y9216a:1926750643) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L436)):*
Pointless `to_vec` call; file a followup (needs an API on `ByteString` ).

---

*[components/script/dom/xmlhttprequest.rs, line 448 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fwR_nWc8eGZlXDxVh:-K9fwR_pPakzlmURlakq:-14134165) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L448)):*
This fix should cause a test to pass; if not, write one.

Also, wrap into three lines.

---

*[components/script/dom/xmlhttprequest.rs, line 451 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fwpAH_58IT-NDjgoU:-K9fwpAJekSspJmrOpZe:658635309) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L451)):*
Add a pointer to the bit of the spec where this is handled.

---

*[components/script/dom/xmlhttprequest.rs, line 483 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fx2Om0DOEc4RO7_og:-K9fx2Om0DOEc4RO7_oh:1282875777) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L483)):*
Split into two separate match arms for symmetry.

---

*[components/script/dom/xmlhttprequest.rs, line 523 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fzfJDHl6zIJs4vP0u:-K9fzfJHzE4yiTq1WaE_:-1039099954) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L523)):*
So we didn't do this in the sync case before; check if there's a way to test that.

---

*[components/script/dom/xmlhttprequest.rs, line 698 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9g1Bq5DgjduwBIHVQO:-K9g1Bq73Ci2Gjy4mzFY:-920266886) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L698)):*
This seems to ignore "excluding any parameters, and converted to ASCII lowercase." Followup.

---

*[components/script/dom/xmlhttprequest.rs, line 702 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9g0s3SQ-6l2YsiA0_W:-K9g0s3VJ6xTwAlOfE1m:-2021788707) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L702)):*
Weird indentation.

---

*[components/script/dom/xmlhttprequest.rs, line 724 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9g1YmxGFfcvddMGuvs:-K9g1YmzjMI_X-MSrgFB:-11440725) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L724)):*
We need to not do this in workers; followup.

Also not sure of the point of doing this in the `match` rather than with a plain `if` below the `match`.

---

*[components/script/dom/xmlhttprequest.rs, line 755 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9g2IGGgzME9mp-2-Is:-K9g2IGIZsBVvHiTzhDX:160941303) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L755)):*
Followup to move this into a function of its own.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383#-:-K9g37Hgi3IJ_vkU2Kd4:-2134094098)*

**Comment by KiChjang** [comment]
Review status: 0 of 1 files reviewed at latest revision, 17 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 308 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fkeam9hi9GbqUTzUC:-K9krRWpN_Zt50uwoFWA:1571469526) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L308)):*
This sounds more like a follow-up issue to fix the `Document` method. I'd say that we should still keep it here since the spec says so.

---

*[components/script/dom/xmlhttprequest.rs, line 416 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fuOpBiiIf0YjTkdU4:-K9kwEmwUS95wSYPdTyF:-796749440) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L416)):*
This should actually be marked `unreachable!()`. The `.as_str()` method returns `Some(&str)` if the `ByteString` is encoded using UTF-8, and `None` otherwise.

---

*[components/script/dom/xmlhttprequest.rs, line 436 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fwCzbNZnmQK79yG4K:-K9kzcfxEeSxHAoKJVkQ:-2082476326) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L436)):*
http://hyper.rs/hyper/hyper/header/struct.Headers.html#method.set_raw seems to accept a `Vec<Vec<u8>>`.

---

*[components/script/dom/xmlhttprequest.rs, line 451 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fwpAH_58IT-NDjgoU:-K9l-e2LSdib1Zprs2V9:171885854) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L451)):*
I don't actually know which part of the spec this algorithm refers to.

---

*[components/script/dom/xmlhttprequest.rs, line 755 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9g2IGGgzME9mp-2-Is:-K9l5nxRjlpth2oeTbqU:-547065674) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L755)):*
Yeah... all of the Response Document/JSON/ArrayBuffer/Blob methods need to be implemented.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383)*

**Comment by KiChjang** [comment]
Review status: 0 of 1 files reviewed at latest revision, 17 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 523 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fzfJDHl6zIJs4vP0u:-K9lARf2EBd16BUXuWzM:905796363) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L523)):*
This can be tested on a few places - calling the `Send` method again for example, or `SetRequestHeader`, `SetWithCredentials` and `Abort`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383)*

**Comment by Ms2ger** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 436 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fwCzbNZnmQK79yG4K:-KAduAdQiGd52OJfHZ_x:-586829673) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L436)):*
Yeah, but we can steal the vec in `value`.

---

*[components/script/dom/xmlhttprequest.rs, line 312 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9383#-KA04X_af8q134nMKHwM:-KA04X_dBvxhQ2DFNPq6:-958206678) ([raw file](https://github.com/servo/servo/blob/de0eb1b08fb48712f4ba357d55002e47d0650b61/components/script/dom/xmlhttprequest.rs#L312)):*
Indentation.

---

*[components/script/dom/xmlhttprequest.rs, line 733 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9383#-KAdvF_qzX0_j3MP9fRB:-KAdvF_qzX0_j3MP9fRC:1511602645) ([raw file](https://github.com/servo/servo/blob/46cd86e451ecaae07419cadf231a2e36d78f1c0b/components/script/dom/xmlhttprequest.rs#L733)):*
Needs a test.

---

*[components/script/dom/xmlhttprequest.rs, line 1105 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9383#-KAdv17gZgzI5tq-KPfP:-KAdv17gZgzI5tq-KPfQ:830245073) ([raw file](https://github.com/servo/servo/blob/46cd86e451ecaae07419cadf231a2e36d78f1c0b/components/script/dom/xmlhttprequest.rs#L1105)):*
Try `Mime::to_string` rather than `ToString::to_string`.

---

*[components/script/dom/xmlhttprequest.rs, line 1107 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9383#-KAdv6b43EIQA5azfMNR:-KAdv6b43EIQA5azfMNS:-1239772751) ([raw file](https://github.com/servo/servo/blob/46cd86e451ecaae07419cadf231a2e36d78f1c0b/components/script/dom/xmlhttprequest.rs#L1107)):*
We don't use `&&` for this anywhere else.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383)*

**Comment by KiChjang** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 436 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fwCzbNZnmQK79yG4K:-KAdyMyRI6en_k0J-sma:-1842944661) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L436)):*
Filed #9655.

---

*[components/script/dom/xmlhttprequest.rs, line 698 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9g1Bq5DgjduwBIHVQO:-KAdynXHqoEKO6AqUyk-:-1842974576) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L698)):*
Filed #9551.

---

*[components/script/dom/xmlhttprequest.rs, line 724 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9g1YmxGFfcvddMGuvs:-KAdzN8JAmZ0588640E9:1780107366) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L724)):*
This codepath shouldn't be possible in a Worker context, because `sync_in_window()` returns false on Worker environments.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383)*

**Comment by KiChjang** [comment]
Review status: 0 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 448 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fwR_nWc8eGZlXDxVh:-KAe5uEYOzYUPRp0wcy3:382606352) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L448)):*
Updated test expections for `xmlhttprequest-timeout-worker-synconworker.htm`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383)*

**Comment by Ms2ger** [comment]
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 451 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fwpAH_58IT-NDjgoU:-KAeJcbifhtr8eqLRALk:-75994435) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L451)):*
Just add an empty line between the code for step 2 and this code, please.

---

*[components/script/dom/xmlhttprequest.rs, line 523 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fzfJDHl6zIJs4vP0u:-KAeNOo1ZSvOwzRhrGGM:-1668191507) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L523)):*
I've tried and failed. @Manishearth, thoughts on ways to test this?

---

*[tests/wpt/web-platform-tests/XMLHttpRequest/responsetype.html, line 87 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9383#-KAeGN072maGPZNK7vQt:-KAeGN072maGPZNK7vQu:625372901) ([raw file](https://github.com/servo/servo/blob/e2eed802c16afb85461fc16211bc60278396d0ed/tests/wpt/web-platform-tests/XMLHttpRequest/responsetype.html#L87)):*
Looks like your new test duplicates this one? I seem to have been confused about what change you actually made.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383)*

**Comment by KiChjang** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[tests/wpt/web-platform-tests/XMLHttpRequest/responsetype.html, line 87 \[r6\]](https://reviewable.io:443/reviews/servo/servo/9383#-KAeGN072maGPZNK7vQt:-KAeOhGNtRDijbpjGe-d:2043649024) ([raw file](https://github.com/servo/servo/blob/e2eed802c16afb85461fc16211bc60278396d0ed/tests/wpt/web-platform-tests/XMLHttpRequest/responsetype.html#L87)):*
Wait, so SetResponseType already has this test to check for an `InvalidAccessError`? Then our test coverage is good enough, I didn't need to write a test.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383)*

**Comment by Manishearth** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 523 \[r3\]](https://reviewable.io:443/reviews/servo/servo/9383#-K9fzfJDHl6zIJs4vP0u:-KAePQjDOIEqxqv4mrjN:-395598011) ([raw file](https://github.com/servo/servo/blob/7c5295eae786291fe6fc3b7db6a4ec5c31775d7c/components/script/dom/xmlhttprequest.rs#L523)):*
Ensure that send() throws an InvalidStateError when called from a readystatechanged handler where  readystate  is headers_received or loading.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383)*

**Comment by Ms2ger** [comment]
@bors-servo r+

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9383#-:-KAoropRjOxh9gQ5Cef5:1236647940)*

**Comment by jdm** [comment]
@bors-servo: retry
* #9488

**Comment by KiChjang** [comment]
@bors-servo retry
* #9092

**Comment by KiChjang** [comment]
@bors-servo retry
* #8120

**Pull request merged**